sky/shell/platform/ios/accessibility_bridge.mm

```diff
@@ -0,0 +1,259 @@
+    : public base::RefCounted<AccessibilityBridge::Node> {
+ public:
+  static const uint32_t kUninitializedNodeId = -1;
+
+  Node(AccessibilityBridge*, const semantics::SemanticsNodePtr&);
+
+  void Update(const semantics::SemanticsNodePtr& node);
+  void PopulateAccessibleElements(NSMutableArray* accessibleElements);
+
+  uint32_t id_ = kUninitializedNodeId;
+  std::vector<scoped_refptr<Node>> children_;
+  Node* parent_ = nullptr;
+
+ private:
+  friend class base::RefCounted<Node>;
+
+  ~Node();
+
+  void ValidateGlobalRect();
+  void ValidateGlobalTransform();
+
+  AccessibilityBridge* bridge_;
+
+  semantics::SemanticFlagsPtr flags_;
+  semantics::SemanticStringsPtr strings_;
+  Geometry geometry_;
+
+  std::unique_ptr<SkMatrix44> global_transform_;
+  std::unique_ptr<SkRect> global_rect_;
+
+  DISALLOW_COPY_AND_ASSIGN(Node);
+};
+
+AccessibilityBridge::Node::Node(AccessibilityBridge* bridge,
+                                const semantics::SemanticsNodePtr& node)
+    : bridge_(bridge) {
+  Update(node);
+}
+
+void AccessibilityBridge::Node::Update(
+    const semantics::SemanticsNodePtr& node) {
+  if (id_ == kUninitializedNodeId) {
+    id_ = node->id;
+  }
+  DCHECK(id_ == node->id);
+
+  if (!node->flags.is_null()) {
+    flags_ = node->flags.Pass();
+  }
+
+  if (!node->strings.is_null()) {
+    strings_ = node->strings.Pass();
+  }
+
+  if (!node->geometry.is_null()) {
+    geometry_ = node->geometry.Pass();
+  }
+
+  if (!node->children.is_null()) {
+    // Mark children for removal
+    for (scoped_refptr<Node> child : children_) {
+      DCHECK(child->parent_ != nullptr);
+      child->parent_ = nullptr;
+    }
+
+    // Set the new list of children
+    std::vector<scoped_refptr<Node>> children;
+    for (const semantics::SemanticsNodePtr& childNode : node->children) {
+      scoped_refptr<Node> child = bridge_->UpdateNode(childNode);
+      child->parent_ = this;
+      children.push_back(child);
+    }
+    children.swap(children_);
+
+    // Remove those children that are still marked for removal
+    for (scoped_refptr<Node> child : children) {
+      if (child->parent_ == nullptr) {
+        bridge_->RemoveNode(child);
+      }
+    }
+  }
+
+  global_transform_.release();
+  global_rect_.release();
+}
+
+void AccessibilityBridge::Node::ValidateGlobalTransform() {
+  if (global_transform_ != nullptr) {
+    return;
+  }
+
+  if (parent_ == nullptr) {
+    global_transform_.reset(new SkMatrix44(geometry_.transform));
+  } else {
+    parent_->ValidateGlobalTransform();
+    const std::unique_ptr<SkMatrix44>& parentTransform =
+        parent_->global_transform_;
+    if (geometry_.transform.isIdentity()) {
+      global_transform_.reset(new SkMatrix44(*parentTransform));
+    } else if (parentTransform->isIdentity()) {
+      global_transform_.reset(new SkMatrix44(geometry_.transform));
+    } else {
+      global_transform_.reset(
+          new SkMatrix44(geometry_.transform * *parentTransform));
+    }
```

Comment:
I wouldn't bother with these special cases.  You can just always perform the matrix multiplication.